Closing intellij during cargo check is running

```
Already disposed: Project (Disposed) analysis
java.lang.AssertionError: Already disposed: Project (Disposed) analysis
	at com.intellij.openapi.components.impl.ComponentManagerImpl.getComponent(ComponentManagerImpl.java:167)
	at com.intellij.psi.PsiManager.getInstance(PsiManager.java:39)
	at org.rust.ide.annotator.CargoCheckAnnotationResult.<init>(RsCargoCheckAnnotator.kt:42)
	at org.rust.ide.annotator.RsCargoCheckAnnotator.checkProject(RsCargoCheckAnnotator.kt:218)
	at org.rust.ide.annotator.RsCargoCheckAnnotator$getCachedResult$1.compute(RsCargoCheckAnnotator.kt:63)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:49)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:222)
	at com.intellij.psi.impl.PsiCachedValue.getValueWithLock(PsiCachedValue.java:60)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:38)
	at org.rust.ide.annotator.RsCargoCheckAnnotator.doAnnotate(RsCargoCheckAnnotator.kt:73)
	at org.rust.ide.annotator.RsCargoCheckAnnotator.doAnnotate(RsCargoCheckAnnotator.kt:58)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.doAnnotate(ExternalToolPass.java:224)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.access$500(ExternalToolPass.java:48)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass$1.run(ExternalToolPass.java:172)
	at com.intellij.util.ui.update.MergingUpdateQueue.execute(MergingUpdateQueue.java:337)
	at com.intellij.util.ui.update.MergingUpdateQueue.execute(MergingUpdateQueue.java:327)
	at com.intellij.util.ui.update.MergingUpdateQueue.lambda$flush$1(MergingUpdateQueue.java:277)
	at com.intellij.util.ui.update.MergingUpdateQueue.flush(MergingUpdateQueue.java:291)
	at com.intellij.util.ui.update.MergingUpdateQueue.run(MergingUpdateQueue.java:246)
	at com.intellij.util.concurrency.QueueProcessor.runSafely(QueueProcessor.java:223)
	at com.intellij.util.Alarm$Request.runSafely(Alarm.java:418)
	at com.intellij.util.Alarm$Request.access$700(Alarm.java:343)
	at com.intellij.util.Alarm$Request$1.run(Alarm.java:385)
	at com.intellij.util.Alarm$Request.run(Alarm.java:396)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at com.intellij.util.concurrency.SchedulingWrapper$MyScheduledFutureTask.run(SchedulingWrapper.java:237)
	at com.intellij.util.concurrency.BoundedTaskExecutor$2.run(BoundedTaskExecutor.java:212)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

--- a/src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt
+++ b/src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt
@@ -36,7 +36,9 @@ import org.rust.ide.RsConstants
 import org.rust.lang.core.psi.ext.module
 import java.util.*
 
-data class CargoCheckAnnotationInfo(val file: PsiFile, val editor: Editor)
+data class CargoCheckAnnotationInfo(
+    val file: PsiFile
+)
 
 class CargoCheckAnnotationResult(commandOutput: List<String>, val project: Project)
     : ModificationTracker by PsiManager.getInstance(project).modificationTracker {
@@ -57,20 +59,18 @@ class CargoCheckAnnotationResult(commandOutput: List<String>, val project: Proje
 
 class RsCargoCheckAnnotator : ExternalAnnotator<CargoCheckAnnotationInfo, CargoCheckAnnotationResult>() {
 
-    private fun getCachedResult(file: PsiFile) =
-        CachedValuesManager.getManager(file.project).createCachedValue {
-            CachedValueProvider.Result.create(
-                checkProject(file),
-                PsiModificationTracker.MODIFICATION_COUNT)
-        }
+    override fun collectInformation(file: PsiFile, editor: Editor, hasErrors: Boolean): CargoCheckAnnotationInfo? {
+        if (!file.project.rustSettings.useCargoCheckAnnotator) return null
 
-    override fun collectInformation(file: PsiFile, editor: Editor, hasErrors: Boolean): CargoCheckAnnotationInfo? =
-        if (file.project.rustSettings.useCargoCheckAnnotator) {
-            CargoCheckAnnotationInfo(file, editor)
-        } else null
+        return CargoCheckAnnotationInfo(file)
+    }
 
     override fun doAnnotate(info: CargoCheckAnnotationInfo): CargoCheckAnnotationResult? =
-        getCachedResult(info.file).value
+        CachedValuesManager.getManager(info.file.project).createCachedValue {
+            CachedValueProvider.Result.create(
+                checkProject(info.file),
+                PsiModificationTracker.MODIFICATION_COUNT)
+        }.value
 
     override fun apply(file: PsiFile, annotationResult: CargoCheckAnnotationResult?, holder: AnnotationHolder) {
         annotationResult ?: return

--- a/src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt
+++ b/src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt
@@ -68,7 +68,7 @@ class RsCargoCheckAnnotator : ExternalAnnotator<CargoCheckAnnotationInfo, CargoC
     override fun doAnnotate(info: CargoCheckAnnotationInfo): CargoCheckAnnotationResult? =
         CachedValuesManager.getManager(info.file.project).createCachedValue {
             CachedValueProvider.Result.create(
-                checkProject(info.file),
+                checkProject(info),
                 PsiModificationTracker.MODIFICATION_COUNT)
         }.value
 
@@ -189,32 +189,33 @@ class RsCargoCheckAnnotator : ExternalAnnotator<CargoCheckAnnotationInfo, CargoC
 
     fun CargoSpan.isValid() = line_end > line_start || (line_end == line_start && column_end >= column_start)
 
-    // NB: executed asynchronously off EDT, so care must be taken not to access
-    // disposed objects
-    fun checkProject(file: PsiFile): CargoCheckAnnotationResult? {
-        val (toolchain, projectPath, disposable: Disposable) = runReadAction {
-            val module = file.module ?: return@runReadAction null
-            if (module.isDisposed) return@runReadAction null
-            val projectRoot = module.cargoProjectRoot ?: return@runReadAction null
-            val toolchain = module.project.toolchain ?: return@runReadAction null
-            Triple(toolchain, projectRoot.path, module)
-        } ?: return null
+}
 
-        // We have to save the file to disk to give cargo a chance to check fresh file content.
-        object : WriteAction<Unit>() {
-            override fun run(result: Result<Unit>) {
-                val fileDocumentManager = FileDocumentManager.getInstance()
-                val document = fileDocumentManager.getDocument(file.virtualFile)
-                if (document == null) {
-                    fileDocumentManager.saveAllDocuments()
-                } else if (fileDocumentManager.isDocumentUnsaved(document)) {
-                    fileDocumentManager.saveDocument(document)
-                }
+// NB: executed asynchronously off EDT, so care must be taken not to access
+// disposed objects
+private fun checkProject(info: CargoCheckAnnotationInfo): CargoCheckAnnotationResult? {
+    val (toolchain, projectPath, disposable: Disposable) = runReadAction {
+        val module = info.file.module ?: return@runReadAction null
+        if (module.isDisposed) return@runReadAction null
+        val projectRoot = module.cargoProjectRoot ?: return@runReadAction null
+        val toolchain = module.project.toolchain ?: return@runReadAction null
+        Triple(toolchain, projectRoot.path, module)
+    } ?: return null
+
+    // We have to save the file to disk to give cargo a chance to check fresh file content.
+    object : WriteAction<Unit>() {
+        override fun run(result: Result<Unit>) {
+            val fileDocumentManager = FileDocumentManager.getInstance()
+            val document = fileDocumentManager.getDocument(info.file.virtualFile)
+            if (document == null) {
+                fileDocumentManager.saveAllDocuments()
+            } else if (fileDocumentManager.isDocumentUnsaved(document)) {
+                fileDocumentManager.saveDocument(document)
             }
-        }.execute()
+        }
+    }.execute()
 
-        val output = toolchain.cargo(projectPath).checkProject(disposable)
-        if (output.isCancelled) return null
-        return CargoCheckAnnotationResult(output.stdoutLines, file.project)
-    }
+    val output = toolchain.cargo(projectPath).checkProject(disposable)
+    if (output.isCancelled) return null
+    return CargoCheckAnnotationResult(output.stdoutLines, info.file.project)
 }

--- a/src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt
+++ b/src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt
@@ -8,18 +8,15 @@ package org.rust.ide.annotator
 import com.google.gson.JsonParser
 import com.intellij.lang.annotation.*
 import com.intellij.lang.annotation.Annotation
-import com.intellij.openapi.Disposable
 import com.intellij.openapi.application.Result
 import com.intellij.openapi.application.WriteAction
-import com.intellij.openapi.application.runReadAction
 import com.intellij.openapi.editor.Document
 import com.intellij.openapi.editor.Editor
 import com.intellij.openapi.fileEditor.FileDocumentManager
-import com.intellij.openapi.project.Project
-import com.intellij.openapi.util.ModificationTracker
+import com.intellij.openapi.module.Module
 import com.intellij.openapi.util.TextRange
+import com.intellij.openapi.vfs.VirtualFile
 import com.intellij.psi.PsiFile
-import com.intellij.psi.PsiManager
 import com.intellij.psi.util.CachedValueProvider
 import com.intellij.psi.util.CachedValuesManager
 import com.intellij.psi.util.PsiModificationTracker
@@ -27,22 +24,20 @@ import com.intellij.util.PathUtil
 import org.apache.commons.lang.StringEscapeUtils.escapeHtml
 import org.rust.cargo.project.settings.rustSettings
 import org.rust.cargo.project.settings.toolchain
-import org.rust.cargo.toolchain.CargoCode
-import org.rust.cargo.toolchain.CargoMessage
-import org.rust.cargo.toolchain.CargoSpan
-import org.rust.cargo.toolchain.CargoTopMessage
+import org.rust.cargo.toolchain.*
 import org.rust.cargo.util.cargoProjectRoot
 import org.rust.ide.RsConstants
 import org.rust.lang.core.psi.ext.module
 import java.util.*
 
 data class CargoCheckAnnotationInfo(
-    val file: PsiFile
+    val file: VirtualFile,
+    val toolchain: RustToolchain,
+    val projectPath: String,
+    val module: Module
 )
 
-class CargoCheckAnnotationResult(commandOutput: List<String>, val project: Project)
-    : ModificationTracker by PsiManager.getInstance(project).modificationTracker {
-
+class CargoCheckAnnotationResult(commandOutput: List<String>) {
     companion object {
         private val parser = JsonParser()
         private val messageRegex = """\s*\{\s*"message".*""".toRegex()
@@ -61,16 +56,20 @@ class RsCargoCheckAnnotator : ExternalAnnotator<CargoCheckAnnotationInfo, CargoC
 
     override fun collectInformation(file: PsiFile, editor: Editor, hasErrors: Boolean): CargoCheckAnnotationInfo? {
         if (!file.project.rustSettings.useCargoCheckAnnotator) return null
+        val module = file.module ?: return null
+        val projectRoot = module.cargoProjectRoot ?: return null
+        val toolchain = module.project.toolchain ?: return null
 
-        return CargoCheckAnnotationInfo(file)
+        return CargoCheckAnnotationInfo(file.virtualFile, toolchain, projectRoot.path, module)
     }
 
     override fun doAnnotate(info: CargoCheckAnnotationInfo): CargoCheckAnnotationResult? =
-        CachedValuesManager.getManager(info.file.project).createCachedValue {
-            CachedValueProvider.Result.create(
-                checkProject(info),
-                PsiModificationTracker.MODIFICATION_COUNT)
-        }.value
+        CachedValuesManager.getManager(info.module.project)
+            .getCachedValue(info.module, {
+                CachedValueProvider.Result.create(
+                    checkProject(info),
+                    PsiModificationTracker.MODIFICATION_COUNT)
+            })
 
     override fun apply(file: PsiFile, annotationResult: CargoCheckAnnotationResult?, holder: AnnotationHolder) {
         annotationResult ?: return
@@ -194,19 +193,11 @@ class RsCargoCheckAnnotator : ExternalAnnotator<CargoCheckAnnotationInfo, CargoC
 // NB: executed asynchronously off EDT, so care must be taken not to access
 // disposed objects
 private fun checkProject(info: CargoCheckAnnotationInfo): CargoCheckAnnotationResult? {
-    val (toolchain, projectPath, disposable: Disposable) = runReadAction {
-        val module = info.file.module ?: return@runReadAction null
-        if (module.isDisposed) return@runReadAction null
-        val projectRoot = module.cargoProjectRoot ?: return@runReadAction null
-        val toolchain = module.project.toolchain ?: return@runReadAction null
-        Triple(toolchain, projectRoot.path, module)
-    } ?: return null
-
     // We have to save the file to disk to give cargo a chance to check fresh file content.
     object : WriteAction<Unit>() {
         override fun run(result: Result<Unit>) {
             val fileDocumentManager = FileDocumentManager.getInstance()
-            val document = fileDocumentManager.getDocument(info.file.virtualFile)
+            val document = fileDocumentManager.getDocument(info.file)
             if (document == null) {
                 fileDocumentManager.saveAllDocuments()
             } else if (fileDocumentManager.isDocumentUnsaved(document)) {
@@ -215,7 +206,7 @@ private fun checkProject(info: CargoCheckAnnotationInfo): CargoCheckAnnotationRe
         }
     }.execute()
 
-    val output = toolchain.cargo(projectPath).checkProject(disposable)
+    val output = info.toolchain.cargo(info.projectPath).checkProject(info.module)
     if (output.isCancelled) return null
-    return CargoCheckAnnotationResult(output.stdoutLines, info.file.project)
+    return CargoCheckAnnotationResult(output.stdoutLines)
 }

